kvpress/prunners/default_pruner.py

```diff
@@ -0,0 +1,92 @@
+# SPDX-License-Identifier: Apache-2.0
+
+
+import logging
+from dataclasses import dataclass
+
+import torch
+from torch import nn
+from transformers import QuantizedCache
+
+from kvpress.prunners.base_pruner import BasePruner
+from kvpress.scorers.base_scorer import BaseScorer
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class DefaultPruner(BasePruner):
+    """
+    Default pruning method.
+    The `forward_hook` method is called after the forward pass of an attention layer.
+    and updates the cache with the pruned KV pairs.
+    """
+
+    scorer: BaseScorer
+    compression_ratio: float = 0.0
+
+    def __post_init__(self):
+        assert 0 <= self.compression_ratio < 1, "Compression ratio must be between 0 and 1"
+
+    def forward_hook(self, module: nn.Module, input: list[torch.Tensor], kwargs: dict, output: list):
+        """
+        Default cache compression hook called after the forward pass of an attention layer.
+        The hook is applied only during the pre-filling phase if there is some pruning ratio.
+        This implementation allows to remove a constant number of KV pairs.
+
+        Parameters
+        ----------
+        module :
+            Transformer attention layer.
+        input :
+            Input to the hook. This is the input to the forward pass of the layer.
+        kwargs :
+            Keyword arguments, as given to the forward pass of the layer.
+        output :
+            Output of the hook. This is the original output of the forward pass of the layer.
+
+        Returns
+        -------
+            Modified output of the forward pass of the layer.
+
+        """
+        # See e.g. LlamaDecoderLayer.forward for the output structure
+        if len(output) == 3:
+            _, attentions, cache = output
+        else:
+            attentions, cache = None, output[-1]
+
+        hidden_states = kwargs["hidden_states"]
+        q_len = hidden_states.shape[1]
+
+        # Don't compress if the compression ratio is 0 or this is not pre-filling
+        if (self.compression_ratio == 0) or (cache.seen_tokens > q_len):
+            return output
+
+        if isinstance(cache, QuantizedCache):
+            keys = cache._dequantize(cache._quantized_key_cache[module.layer_idx])
+            values = cache._dequantize(cache._quantized_value_cache[module.layer_idx])
+        else:
+            keys = cache.key_cache[module.layer_idx]
+            values = cache.value_cache[module.layer_idx]
+
+        with torch.no_grad():
+            scores = self.scorer.score(module, hidden_states, keys, values, attentions, kwargs)
+
+        # Prune KV pairs with the lowest scores
+        n_kept = int(q_len * (1 - self.compression_ratio))
+        indices = scores.topk(n_kept, dim=-1).indices
+        indices = indices.unsqueeze(-1).expand(-1, -1, -1, module.head_dim)
+
+        # Update cache
+        keys = keys.gather(2, indices).contiguous()
+        values = values.gather(2, indices).contiguous()
```

Comment:
Only these lines are specific to a scorer. All the other lines (sanity checks, quantize and dequantize) are generic. It could be nice to create a new `prune` method for these lines only. It would simplify the forward_hook of other presses, such as ThinKPress and SimLayerKVPress were we needed to re-rewrite the sanity checks + quantization / dequantization lines.

Reply:
also I believe you can remove the torch.no_grad

Reply:
based on this comment, I don't think we really need a `Scorer` class anymore. I like the idea of separating presses with a very clear logic. But the only place where the `Scorer` is used is in `DefaultPruner` so all "scorers" could simply inherit from it and we could rename the directory `score_presses` and `DefaultPruner` to `ScorePress` or something like this

Reply:
> maybe we could move this to the BasePress. As before the refacto the BasePress has a forward_hook method but in the middle of this foward hook there is this compress method which raises not ImplementedError. 

I'd leave the base press' (base pruner) `forward_hook` completely empty for now to have a very slim base class.
The code duplication is minor atm. We can refactor if we see more code duplication appear.

Reply:
There's already duplicated code in ThinK, and soon in SimLayerKV. Anyone willing to implement a press will need to copy past those lines.

Reply:
That being said you can let it to me and I'll implement it for simlayerkv PR